In [2]:
from multiprocessing import Pool, cpu_count
import gc; gc.enable()
import xgboost as xgb
import pandas as pd
import numpy as np
from sklearn import *
import sklearn
from sklearn.model_selection import cross_val_score, GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report,confusion_matrix
from sklearn.model_selection import RandomizedSearchCV
from sklearn.linear_model import LogisticRegression

In [77]:
# reading wsdm csv
wsdm_data = pd.read_csv(r"D:\Work_Masterarbeit\wsdm\train_v2.csv")

# reading members csv
members_data = pd.read_csv(r"D:\Work_Masterarbeit\wsdm\members_v3.csv")

# merging train dataset with members dataset
wsdm_members = pd.merge(wsdm_data, members_data, on='msno', how='left')
# getting the head (top 5 rows) of df
wsdm_members.head()

msno  is_churn  city    bd  gender  \
0  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0    male   
1  f/NmvEzHfhINFEYZTR05prUdr+E+3+oewvweYz9cCQE=         1  13.0  20.0    male   
2  zLo9f73nGGT1p21ltZC3ChiRnAVvgibMyazbCxvWPcg=         1  13.0  18.0    male   
3  8iF/+8HY8lJKFrTc7iR9ZYGCG2Ecrogbc2Vy5YhsfhQ=         1   1.0   0.0     NaN   
4  K6fja4+jmoZ5xG6BypqX80Uw/XKpMgrEMdG2edFOxnA=         1  13.0  35.0  female   

   registered_via  registration_init_time  
0             3.0              20131223.0  
1             3.0              20131223.0  
2             3.0              20131227.0  
3             7.0              20140109.0  
4             7.0              20140125.0

In [78]:
### 4.1 Loading files to merge-

In [79]:
# Laden der Daten wsdm Daten [VOV]
wsdm_data = pd.read_csv(r"D:\Work_Masterarbeit\wsdm\train_v2.csv")
# members data
members_data = pd.read_csv(r"D:\Work_Masterarbeit\wsdm\members_v3.csv")
# transactions data
transactions_data = pd.read_csv(r"D:\Work_Masterarbeit\wsdm\transactions_v2.csv")
# user logs data
logs_data = pd.read_csv(r"D:\Work_Masterarbeit\wsdm\user_logs_v2.csv")

In [80]:
### 4.2 Merging files together-
### 4.2.1 For train data-

In [81]:
# Zusammenführen der Daten [VOV]
wsdm_members = pd.merge(wsdm_data, members_data, on='msno', how='left')
# merging transactions data with train members data
wsdm_mem_trans = pd.merge(wsdm_members, transactions_data, on='msno', how='left')
# merging user logs data with train members transaction data
wsdm_dataset = pd.merge(wsdm_mem_trans, logs_data, on='msno', how='left')
# getting head of train dataset
pd.set_option('display.max_columns', 100)
wsdm_dataset.head()

msno  is_churn  city    bd gender  \
0  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0   male   
1  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0   male   
2  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0   male   
3  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0   male   
4  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0   male   

   registered_via  registration_init_time  payment_method_id  \
0             3.0              20131223.0                NaN   
1             3.0              20131223.0                NaN   
2             3.0              20131223.0                NaN   
3             3.0              20131223.0                NaN   
4             3.0              20131223.0                NaN   

   payment_plan_days  plan_list_price  actual_amount_paid  is_auto_renew  \
0                NaN              NaN                 NaN            NaN   
1                NaN              NaN                 NaN            NaN   
2                NaN              NaN                 NaN            NaN   
3                NaN              NaN                 NaN            NaN   
4                NaN              NaN                 NaN            NaN   

   transaction_date  membership_expire_date  is_cancel        date  num_25  \
0               NaN                     NaN        NaN  20170305.0     7.0   
1               NaN                     NaN        NaN  20170301.0   138.0   
2               NaN                     NaN        NaN  20170319.0     0.0   
3               NaN                     NaN        NaN  20170316.0    15.0   
4               NaN                     NaN        NaN  20170310.0     0.0   

   num_50  num_75  num_985  num_100  num_unq  total_secs  
0     0.0     3.0      5.0     71.0     68.0   17599.893  
1    19.0     7.0      1.0     21.0    158.0    8830.433  
2     0.0     0.0      0.0     34.0     17.0    7883.313  
3     0.0     0.0      1.0     38.0     17.0    9029.227  
4     0.0     0.0      0.0      8.0      8.0    1870.110

In [82]:
wsdm_dataset.shape

(16887877, 23)

In [83]:
# [VOV] since registration_init_time contains a lot of nan values so impute median date
wsdm_dataset['registration_init_time'] = wsdm_dataset['registration_init_time'].fillna(20150204.0)
# getting registration_init_time array
reg_date = wsdm_dataset['registration_init_time'].values
reg_date

array([20131223., 20131223., 20131223., ..., 20110905., 20110905.,
       20150204.])

In [84]:
# storing these train registrationn dates to npz file
np.savez_compressed(r"D:\Work_Masterarbeit\wsdm\preprocessed\wsdm_registration_date", reg_date)

<h3> Vorverarbeitung
   

In [85]:
def preprocess(data):
    '''[VOV] Diese Vorverarbeitungsfunktion wird verwendet, um grundlegende Vorverarbeitungen am Anfang des 
    Trainings- und Testdatensatzes durchzuführen. Diese Vorverarbeitung besteht aus der Entfernung von Ausreißern, 
    der Imputation von Null Values und dem Ersetzen der Werte'''
    
    # imputing 0 in place of nan values in the city column
    data['city'] = data['city'].fillna(0)

    # removing outliers
    data['bd'] = data['bd'].apply(lambda x: x if (x < 72.0) and (x > 0.0) else np.nan)
    # imputing 28 as age instead of nan
    data['bd'] = data['bd'].fillna(28.0)

    # replacing male with 1 in gender
    data['gender'] = data['gender'].replace(to_replace='male', value=1)
    # replacing male with 2 in gender
    data['gender'] = data['gender'].replace(to_replace='female', value=2)
    # replacing nan with 0 in gender
    data['gender'] = data['gender'].fillna(0)

    # replace 0 instead of nan in registered_via
    data['registered_via'] = data['registered_via'].fillna(0)

    # filling median date in place of nan in the df
    data['registration_init_time'] = data['registration_init_time'].fillna(20150204.0)
    # converting float date to datetime
    data['registration_init_time'] = pd.to_datetime(data['registration_init_time'], format='%Y%m%d')

    # imputing 0 in place of nan value in payment_method_id
    data['payment_method_id'] = data['payment_method_id'].fillna(0)

    # removing outliers
    data['payment_plan_days'] = data['payment_plan_days'].apply(lambda x: x if (x <= 30.0) else np.nan)
    # imputing 30 in place of nan in payment_plan_days
    data['payment_plan_days'] = data['payment_plan_days'].fillna(30.0)

    # removing outliers
    data['plan_list_price'] = data['plan_list_price'].apply(lambda x: x if (x <= 180.0) else np.nan)
    # imputing 149 in place of nan in plan_list_price
    data['plan_list_price'] = data['plan_list_price'].fillna(149.0)

    # removing outliers
    data['actual_amount_paid'] = data['actual_amount_paid'].apply(lambda x: x if (x <= 180.0) else np.nan)
    # imputing 149 in place of nan in actual_amount_paid
    data['actual_amount_paid'] = data['actual_amount_paid'].fillna(149.0)

    # imputing 2 in place of nan values in is_auto_renew
    data['is_auto_renew'] = data['is_auto_renew'].fillna(2)

    # filling median date in place of nan in the df
    data['transaction_date'] = data['transaction_date'].fillna(20170316.0)
    # converting float date to datetime
    data['transaction_date'] = pd.to_datetime(data['transaction_date'], format='%Y%m%d')

    # filling median date in place of nan in the df
    data['membership_expire_date'] = data['membership_expire_date'].fillna(20170419.0)
    # converting float date to datetime
    data['membership_expire_date'] = pd.to_datetime(data['membership_expire_date'], format='%Y%m%d')

    # imputing 2 in place of nan values in is_cancel
    data['is_cancel'] = data['is_cancel'].fillna(2)

    # filling median date in place of nan in the df
    data['date'] = data['date'].fillna(20170316.0)
    # converting float date to datetime
    data['date'] = pd.to_datetime(data['date'], format='%Y%m%d')
    
    # removing outliers
    data['num_25'] = data['num_25'].apply(lambda x: x if (x <= 15.0) else np.nan)
    # now I can impute 2 instead of nan in num_25
    data['num_25'] = data['num_25'].fillna(2.0)

    # removing outliers
    data['num_50'] = data['num_50'].apply(lambda x: x if (x <= 4.0) else np.nan)
    # now I can impute 0 instead of nan in num_50
    data['num_50'] = data['num_50'].fillna(0)

    # removing outliers
    data['num_75'] = data['num_75'].apply(lambda x: x if (x <= 3.0) else np.nan)
    # now I can impute 0 instead of nan in num_75
    data['num_75'] = data['num_75'].fillna(0)

    # removing outliers
    data['num_985'] = data['num_985'].apply(lambda x: x if (x <= 3.0) else np.nan)
    # now I can impute 0 instead of nan in num_985
    data['num_985'] = data['num_985'].fillna(0)

    # removing outliers
    data['num_100'] = data['num_100'].apply(lambda x: x if (x <= 74.0) else np.nan)
    # now I can impute 14 instead of nan in num_100
    data['num_100'] = data['num_100'].fillna(14.0)

    # removing outliers
    data['num_unq'] = data['num_unq'].apply(lambda x: x if (x <= 68.0) else np.nan)
    # now I can impute 16 instead of nan in num_unq
    data['num_unq'] = data['num_unq'].fillna(16.0)

    # removing outliers
    data['total_secs'] = data['total_secs'].apply(lambda x: x if (x <= 19167.549700000025) else np.nan)
    # now I can impute 3880.765 instead of nan in total_secs
    data['total_secs'] = data['total_secs'].fillna(3880.765)

In [86]:
# Checking null values for train data
# getting features wise null values (number)
wsdm_dataset.isnull().sum()

msno                            0
is_churn                        0
city                       117166
bd                         117166
gender                    7860346
registered_via             117166
registration_init_time          0
payment_method_id          340641
payment_plan_days          340641
plan_list_price            340641
actual_amount_paid         340641
is_auto_renew              340641
transaction_date           340641
membership_expire_date     340641
is_cancel                  340641
date                       249615
num_25                     249615
num_50                     249615
num_75                     249615
num_985                    249615
num_100                    249615
num_unq                    249615
total_secs                 249615
dtype: int64

In [87]:
# performing preprocessing on top of entire train data
preprocess(wsdm_dataset)

In [88]:
# again just to cross check if there is any null or not
wsdm_dataset.isnull().values.any()

False

In [89]:
# checking for any duplicate row
wsdm_dataset.duplicated().any()

False

In [90]:
# saving train file to disk
wsdm_dataset.to_csv(r"D:\Work_Masterarbeit\wsdm\preprocessed\train.csv", index=False)

<h2>Feature Engineering

In [91]:
# loading the files

wsdm_dataset = pd.read_csv(r"D:\Work_Masterarbeit\wsdm\preprocessed\train.csv")

In [92]:
# getting head of wsdm file
pd.set_option('display.max_columns', 100)
wsdm_dataset.head()

msno  is_churn  city    bd  gender  \
0  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   
1  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   
2  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   
3  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   
4  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   

   registered_via registration_init_time  payment_method_id  \
0             3.0             2013-12-23                0.0   
1             3.0             2013-12-23                0.0   
2             3.0             2013-12-23                0.0   
3             3.0             2013-12-23                0.0   
4             3.0             2013-12-23                0.0   

   payment_plan_days  plan_list_price  actual_amount_paid  is_auto_renew  \
0               30.0            149.0               149.0            2.0   
1               30.0            149.0               149.0            2.0   
2               30.0            149.0               149.0            2.0   
3               30.0            149.0               149.0            2.0   
4               30.0            149.0               149.0            2.0   

  transaction_date membership_expire_date  is_cancel        date  num_25  \
0       2017-03-16             2017-04-19        2.0  2017-03-05     7.0   
1       2017-03-16             2017-04-19        2.0  2017-03-01     2.0   
2       2017-03-16             2017-04-19        2.0  2017-03-19     0.0   
3       2017-03-16             2017-04-19        2.0  2017-03-16    15.0   
4       2017-03-16             2017-04-19        2.0  2017-03-10     0.0   

   num_50  num_75  num_985  num_100  num_unq  total_secs  
0     0.0     3.0      0.0     71.0     68.0   17599.893  
1     0.0     0.0      1.0     21.0     16.0    8830.433  
2     0.0     0.0      0.0     34.0     17.0    7883.313  
3     0.0     0.0      1.0     38.0     17.0    9029.227  
4     0.0     0.0      0.0      8.0      8.0    1870.110

<h2> Hier Baseline für ADL

In [93]:
# Feature - 1 (getting weekday from date)

# getting weekday from the date feature for train data
wsdm_dataset['date'] = pd.to_datetime(wsdm_dataset['date'], errors='coerce')
wsdm_dataset['day_of_the_week'] = wsdm_dataset['date'].dt.day_name().values


In [94]:
# Feature - 2 (checking for the weekend)

wsdm_dataset['is_weekend'] = wsdm_dataset['day_of_the_week'].apply(lambda x: 1 if (x == 'Saturday') or (x == 'Sunday') else 0)

In [95]:
# Feature - 3 (checking for the weekday)

wsdm_dataset['is_weekday'] = wsdm_dataset['is_weekend'].apply(lambda x: 1 if (x == 0) else 0)


In [96]:
%%time
# Feature - 4 to 12 (sum based features)
# Feature - 13 to 19 (mean based features)
# Feature - 20 to 27 (standard deviation based features)
# Feature - 28 (nunique based feature)
# Feature - 29 and 30 (min and max based features)
# Feature - 31 to 33 (mean based features for transaction)
# Feature - 34 (transaction count)
# Feature - 35 (transaction date max)
# Feature - 36 (membership expiry date max)
# Feature - 37 (membership expiry date count)

def std(x):
    '''finding standard deviation using numpy,
    to avoid getting nan values'''
    return np.std(x)

# grouping them together for train data
temp_df_train = wsdm_dataset.groupby('msno').agg(num_25_sum=('num_25', 'sum'),
                                num_50_sum=('num_50', 'sum'),
                                num_75_sum=('num_75', 'sum'),
                                num_985_sum=('num_985', 'sum'),
                                num_100_sum=('num_100', 'sum'),
                                num_unq_sum=('num_unq', 'sum'),
                                total_secs_sum=('total_secs', 'sum'),
                                is_weekend_sum=('is_weekend', 'sum'),
                                is_weekday_sum=('is_weekday', 'sum'),
                                num_25_mean=('num_25', 'mean'),
                                num_50_mean=('num_50', 'mean'),
                                num_75_mean=('num_75', 'mean'),
                                num_985_mean=('num_985', 'mean'),
                                num_100_mean=('num_100', 'mean'),
                                num_unq_mean=('num_unq', 'mean'),
                                total_secs_mean=('total_secs', 'mean'),
                                num_25_std=('num_25', std),
                                num_50_std=('num_50', std),
                                num_75_std=('num_75', std),
                                num_985_std=('num_985', std),
                                num_100_std=('num_100', std),
                                num_unq_std=('num_unq', std),
                                total_secs_std=('total_secs', std),
                                active_days=('date', 'nunique'),
                                date_min=('date', 'min'),
                                date_max=('date', 'max'),
                                payment_plan_days_mean=('payment_plan_days', 'mean'),
                                plan_list_price_mean=('plan_list_price', 'mean'),
                                actual_amount_paid_mean=('actual_amount_paid', 'mean'),
                                transaction_date_count=('transaction_date', 'nunique'),
                                transaction_date_max=('transaction_date', 'max'),
                                membership_expire_date_max=('membership_expire_date', 'max'),
                                membership_expire_count=('membership_expire_date', 'nunique'))
# merging them with the train dataset
wsdm_dataset = pd.merge(wsdm_dataset, temp_df_train, on='msno', how='left')

Wall time: 5min 9s


In [97]:
# Feature - 38 (activity period)

wsdm_dataset['date_min'] = pd.to_datetime(wsdm_dataset['date_min'], errors='coerce')
wsdm_dataset['date_max'] = pd.to_datetime(wsdm_dataset['date_max'], errors='coerce')


wsdm_dataset['activity_period'] = (wsdm_dataset['date_max'] - wsdm_dataset['date_min']).dt.days + 1

In [98]:
# Feature - 39 (inactive days)

wsdm_dataset['inactive_days'] = wsdm_dataset['date'].nunique() - wsdm_dataset['active_days']

In [99]:
# Feature - 40 (rare behaviour)
# 0 (for not rare user) and 1 (for rare user)

wsdm_dataset['is_rare'] = wsdm_dataset['active_days'].apply(lambda x: 0 if (x > 1) else 1)

In [100]:
# Feature - 41 (average time per day)

wsdm_dataset['avg_time_perday'] = wsdm_dataset['total_secs_sum'] / wsdm_dataset['active_days']


# Feature - 42 (unique tracks played per day)

wsdm_dataset['unq_track_perday'] = round(wsdm_dataset['num_unq_sum'] / wsdm_dataset['active_days'])


# Feature - 43 (tracks played till 25% length per day)

wsdm_dataset['till_25_perday'] = round(wsdm_dataset['num_25_sum'] / wsdm_dataset['active_days'])


# Feature - 44 (tracks played till 50% length per day)

wsdm_dataset['till_50_perday'] = round(wsdm_dataset['num_50_sum'] / wsdm_dataset['active_days'])


# Feature - 45 (tracks played till 75% length per day)

wsdm_dataset['till_75_perday'] = round(wsdm_dataset['num_75_sum'] / wsdm_dataset['active_days'])


# Feature - 46 (tracks played till 98.5% length per day)

wsdm_dataset['till_985_perday'] = round(wsdm_dataset['num_985_sum'] / wsdm_dataset['active_days'])


# Feature - 47 (tracks played till 100% length per day)

wsdm_dataset['till_full_perday'] = round(wsdm_dataset['num_100_sum'] / wsdm_dataset['active_days'])

In [101]:
# Feature - 48 (discount)
wsdm_dataset['discount'] = wsdm_dataset['plan_list_price'] - wsdm_dataset['actual_amount_paid']

In [102]:
wsdm_dataset['discount'].unique()

array([   0.,  149.,  180., -149.,   30., -119., -129.,  120.,   20.,
         50., -100.,   -1.,   99.,  129.,    1.])

In [103]:
# since there are several values which are less than 0 for various reasons
# so it's always good to set a lower limit, here I can set 0
wsdm_dataset['discount'] = wsdm_dataset['discount'].clip(lower=0)

In [104]:
# Feature - 49 (is_discount)
wsdm_dataset['is_discount'] = wsdm_dataset['discount'].apply(lambda x: 1 if (x > 0) else 0)

In [105]:
# Feature - 50 (days since final login)
# Threshold that I set here is the last date, which is 31-03-2017

wsdm_dataset['days_since_last_login'] = (pd.to_datetime(wsdm_dataset['date'].unique().max(), errors='coerce') - wsdm_dataset['date_max']).dt.days

In [106]:
wsdm_dataset['membership_expire_date_max'] = pd.to_datetime(wsdm_dataset['membership_expire_date_max'], errors='coerce')

wsdm_dataset['days_left'] = (wsdm_dataset['membership_expire_date_max'] - pd.to_datetime(wsdm_dataset['date'].unique().max(), errors='coerce')).dt.days

In [107]:
# Since there are many negative values so I can set them to 0
wsdm_dataset['days_left'] = wsdm_dataset['days_left'].clip(lower=0)

In [108]:
# Feature - 51 (Loyality range)

wsdm_dataset['transaction_date_max'] = pd.to_datetime(wsdm_dataset['transaction_date_max'], errors='coerce')
wsdm_dataset['registration_init_time'] = pd.to_datetime(wsdm_dataset['registration_init_time'], errors='coerce')

wsdm_dataset['layality_range'] = (wsdm_dataset['transaction_date_max'] - wsdm_dataset['registration_init_time']).dt.days

In [109]:
# Feature - 52 (price per day)

wsdm_dataset['Perday_price'] = wsdm_dataset['actual_amount_paid'] / wsdm_dataset['payment_plan_days']

In [110]:
# Feature - 53 (days since final transaction)

wsdm_dataset['days_since_final_transaction'] = (pd.to_datetime(wsdm_dataset['date'].unique().max(), errors='coerce') - wsdm_dataset['transaction_date_max']).dt.days

In [111]:
# all features in train data
wsdm_dataset.columns

Index(['msno', 'is_churn', 'city', 'bd', 'gender', 'registered_via',
       'registration_init_time', 'payment_method_id', 'payment_plan_days',
       'plan_list_price', 'actual_amount_paid', 'is_auto_renew',
       'transaction_date', 'membership_expire_date', 'is_cancel', 'date',
       'num_25', 'num_50', 'num_75', 'num_985', 'num_100', 'num_unq',
       'total_secs', 'day_of_the_week', 'is_weekend', 'is_weekday',
       'num_25_sum', 'num_50_sum', 'num_75_sum', 'num_985_sum', 'num_100_sum',
       'num_unq_sum', 'total_secs_sum', 'is_weekend_sum', 'is_weekday_sum',
       'num_25_mean', 'num_50_mean', 'num_75_mean', 'num_985_mean',
       'num_100_mean', 'num_unq_mean', 'total_secs_mean', 'num_25_std',
       'num_50_std', 'num_75_std', 'num_985_std', 'num_100_std', 'num_unq_std',
       'total_secs_std', 'active_days', 'date_min', 'date_max',
       'payment_plan_days_mean', 'plan_list_price_mean',
       'actual_amount_paid_mean', 'transaction_date_count',
       'transaction_dat

In [112]:
# reference: https://medium.com/@vincentteyssier/optimizing-the-size-of-a-pandas-dataframe-for-low-memory-environment-5f07db3d72e
# reference: https://numpy.org/doc/stable/reference/generated/numpy.iinfo.html
# reference: https://numpy.org/doc/stable/reference/generated/numpy.finfo.html

def datatype_changer(dataset):
    # iterating through all the columns in the dataframe
    for col in dataset.columns:
        # getting column's datatype
        col_type = dataset[col].dtype
        
        # checking if datatype of column is 'object' or not
        # if column type is not object
        if (col_type == int) or (col_type == float):
            # getting minimum value of a column
            min_val = dataset[col].min()
            # getting maximum value of a column
            max_val = dataset[col].max()
            # checking whether the datatype contain first 3 characters as int or not, if int
            if str(col_type)[:3] == 'int':
                # cheking the minimal and maximal value for int8, int16, int32 and int64 in numpy
                if min_val > np.iinfo(np.int8).min and max_val < np.iinfo(np.int8).max:
                    dataset[col] = dataset[col].astype(np.int8)
                elif min_val > np.iinfo(np.int16).min and max_val < np.iinfo(np.int16).max:
                    dataset[col] = dataset[col].astype(np.int16)
                elif min_val > np.iinfo(np.int32).min and max_val < np.iinfo(np.int32).max:
                    dataset[col] = dataset[col].astype(np.int32)
                else:
                    dataset[col] = dataset[col].astype(np.int64)
            else:
                # if it is non int, which is ultimately float
                # cheking the minimal and maximal value for float16, float32 and float64 in numpy
                if min_val > np.finfo(np.float16).min and max_val < np.finfo(np.float16).max:
                    dataset[col] = dataset[col].astype(np.float16)
                elif min_val > np.finfo(np.float32).min and max_val < np.finfo(np.float32).max:
                    dataset[col] = dataset[col].astype(np.float32)
                else:
                    dataset[col] = dataset[col].astype(np.float64)
        else:
            # keeping rest of them to category datatype instead of object
            dataset[col] = dataset[col].astype('category')
            
    # returning head of the dataframe
    return dataset

In [113]:
wsdm_dataset = datatype_changer(wsdm_dataset)

In [114]:
wsdm_dataset

msno is_churn  city    bd  \
0         ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=        1   5.0  28.0   
1         ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=        1   5.0  28.0   
2         ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=        1   5.0  28.0   
3         ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=        1   5.0  28.0   
4         ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=        1   5.0  28.0   
...                                                ...      ...   ...   ...   
16887872  iZE41tbAQ65rJq60olkJT4BJzuUAYgQdfbEemXe/TTk=        0   5.0  25.0   
16887873  iZE41tbAQ65rJq60olkJT4BJzuUAYgQdfbEemXe/TTk=        0   5.0  25.0   
16887874  iZE41tbAQ65rJq60olkJT4BJzuUAYgQdfbEemXe/TTk=        0   5.0  25.0   
16887875  iZE41tbAQ65rJq60olkJT4BJzuUAYgQdfbEemXe/TTk=        0   5.0  25.0   
16887876  oECkzJik4wKsbOEVY6UACLbmgM8qymFdb5cJaHrodY8=        0   0.0  28.0   

          gender  registered_via registration_init_time  payment_method_id  \
0            1.0             3.0             2013-12-23                0.0   
1            1.0             3.0             2013-12-23                0.0   
2            1.0             3.0             2013-12-23                0.0   
3            1.0             3.0             2013-12-23                0.0   
4            1.0             3.0             2013-12-23                0.0   
...          ...             ...                    ...                ...   
16887872     2.0             9.0             2011-09-05               39.0   
16887873     2.0             9.0             2011-09-05               39.0   
16887874     2.0             9.0             2011-09-05               39.0   
16887875     2.0             9.0             2011-09-05               39.0   
16887876     0.0             0.0             2015-02-04               37.0   

          payment_plan_days  plan_list_price  actual_amount_paid  \
0                      30.0            149.0               149.0   
1                      30.0            149.0               149.0   
2                      30.0            149.0               149.0   
3                      30.0            149.0               149.0   
4                      30.0            149.0               149.0   
...                     ...              ...                 ...   
16887872               30.0            149.0               149.0   
16887873               30.0            149.0               149.0   
16887874               30.0            149.0               149.0   
16887875               30.0            149.0               149.0   
16887876               30.0            149.0               149.0   

          is_auto_renew transaction_date membership_expire_date  is_cancel  \
0                   2.0       2017-03-16             2017-04-19        2.0   
1                   2.0       2017-03-16             2017-04-19        2.0   
2                   2.0       2017-03-16             2017-04-19        2.0   
3                   2.0       2017-03-16             2017-04-19        2.0   
4                   2.0       2017-03-16             2017-04-19        2.0   
...                 ...              ...                    ...        ...   
16887872            1.0       2017-02-28             2017-04-13        0.0   
16887873            1.0       2017-02-28             2017-04-13        0.0   
16887874            1.0       2017-02-28             2017-04-13        0.0   
16887875            1.0       2017-02-28             2017-04-13        0.0   
16887876            1.0       2017-03-18             2017-04-18        0.0   

               date  num_25  num_50  num_75  num_985  num_100  num_unq  \
0        2017-03-05     7.0     0.0     3.0      0.0     71.0     68.0   
1        2017-03-01     2.0     0.0     0.0      1.0     21.0     16.0   
2        2017-03-19     0.0     0.0     0.0      0.0     34.0     17.0   
3        2017-03-16    15.0     0.0     0.0      1.0     38.0     17.0   
4        2017-03-10     0.0     0.0     

In [115]:
# saving train file

wsdm_dataset.to_csv(r"D:\Work_Masterarbeit\wsdm\preprocessed\wsdm_FA.csv", index=False)

<h1>3. Prepare Data for Modelling

In [2]:
# loading train dataset
wsdm_dataset = pd.read_csv(r"D:\Work_Masterarbeit\wsdm\preprocessed\wsdm_FA.csv")
# setting max_columns to 100, so it displays all the columns of train dataset
pd.set_option('display.max_columns', 100)
# displaying head (top 5) rows of train dataset
wsdm_dataset.head()

msno  is_churn  city    bd  gender  \
0  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   
1  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   
2  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   
3  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   
4  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   

   registered_via  registration_init_time  payment_method_id  \
0             3.0     1387756800000000000                0.0   
1             3.0     1387756800000000000                0.0   
2             3.0     1387756800000000000                0.0   
3             3.0     1387756800000000000                0.0   
4             3.0     1387756800000000000                0.0   

   payment_plan_days  plan_list_price  actual_amount_paid  is_auto_renew  \
0               30.0            149.0               149.0            2.0   
1               30.0            149.0               149.0            2.0   
2               30.0            149.0               149.0            2.0   
3               30.0            149.0               149.0            2.0   
4               30.0            149.0               149.0            2.0   

  transaction_date membership_expire_date  is_cancel                 date  \
0       2017-03-16             2017-04-19        2.0  1488672000000000000   
1       2017-03-16             2017-04-19        2.0  1488326400000000000   
2       2017-03-16             2017-04-19        2.0  1489881600000000000   
3       2017-03-16             2017-04-19        2.0  1489622400000000000   
4       2017-03-16             2017-04-19        2.0  1489104000000000000   

   num_25  num_50  num_75  num_985  num_100  num_unq  total_secs  \
0     7.0     0.0     3.0      0.0     71.0     68.0     17600.0   
1     2.0     0.0     0.0      1.0     21.0     16.0      8830.0   
2     0.0     0.0     0.0      0.0     34.0     17.0      7884.0   
3    15.0     0.0     0.0      1.0     38.0     17.0      9030.0   
4     0.0     0.0     0.0      0.0      8.0      8.0      1870.0   

  day_of_the_week  is_weekend  is_weekday  num_25_sum  num_50_sum  num_75_sum  \
0          Sunday           1           0        50.0         4.0         6.0   
1       Wednesday           0           1        50.0         4.0         6.0   
2          Sunday           1           0        50.0         4.0         6.0   
3        Thursday           0           1        50.0         4.0         6.0   
4          Friday           0           1        50.0         4.0         6.0   

   num_985_sum  num_100_sum  num_unq_sum  total_secs_sum  is_weekend_sum  \
0          5.0        246.0        206.0       64499.977               5   
1          5.0        246.0        206.0       64499.977               5   
2          5.0        246.0        206.0       64499.977               5   
3          5.0        246.0        206.0       64499.977               5   
4          5.0        246.0        206.0       64499.977               5   

   is_weekday_sum  num_25_mean  num_50_mean  num_75_mean  num_985_mean  \
0               6        4.547       0.3635       0.5454        0.4546   
1               6        4.547       0.3635       0.5454        0.4546   
2               6        4.547       0.3635       0.5454        0.4546   
3               6        4.547       0.3635       0.5454        0.4546   
4               6        4.547       0.3635       0.5454        0.4546   

   num_100_mean  num_unq_mean  total_secs_mean  num_25_std  num_50_std  \
0         22.36         18.73           5864.0       4.734       0.481   
1         22.36         18.73           5864.0       4.734       0.481   
2         22.36         18.73           5864.0       4.734       0.481   
3         22.36         18.73           5864.0       4.734       0.481   
4         22.36         18.73           5864.0       4.734       0.481   

   num_7

In [3]:
# removing all non_useful features first
wsdm_dataset = wsdm_dataset.drop(['registration_init_time',
                              'transaction_date', 
                              'membership_expire_date', 
                              'date', 
                              'num_25',
                              'num_50',
                              'num_75',
                              'num_985',
                              'num_100',
                              'num_unq',
                              'total_secs',
                              'is_weekend',
                              'is_weekday',
                              'day_of_the_week', 
                              'date_min', 
                              'date_max', 
                              'transaction_date_max', 
                              'membership_expire_date_max'], axis=1, inplace=False)

In [4]:
# getting shape of train dataset
wsdm_dataset.shape

(16887877, 58)

Now this train dataset has 16887877 rows and 58 columns

In [5]:
# adding registration init time feature to train dataset
# load the registration array
reg_date = np.load(r"D:\Work_Masterarbeit\wsdm\preprocessed\wsdm_registration_date.npz")
# display the array
reg_date['arr_0']

array([20131223., 20131223., 20131223., ..., 20110905., 20110905.,
       20150204.])

In [6]:
# converting numpy array to list
registration_init_time = reg_date['arr_0'].tolist()
# getting length of the list
len(registration_init_time)

16887877

In [7]:
# creating new feature, and adding it to train dataset
wsdm_dataset['registration_init_time'] = registration_init_time

In [8]:
# setting max_columns to 100, so it displays all the columns of train dataset
pd.set_option('display.max_columns', 100)
# displaying head (top 5) rows of train dataset
wsdm_dataset.head()

msno  is_churn  city    bd  gender  \
0  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   
1  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   
2  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   
3  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   
4  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   

   registered_via  payment_method_id  payment_plan_days  plan_list_price  \
0             3.0                0.0               30.0            149.0   
1             3.0                0.0               30.0            149.0   
2             3.0                0.0               30.0            149.0   
3             3.0                0.0               30.0            149.0   
4             3.0                0.0               30.0            149.0   

   actual_amount_paid  is_auto_renew  is_cancel  num_25_sum  num_50_sum  \
0               149.0            2.0        2.0        50.0         4.0   
1               149.0            2.0        2.0        50.0         4.0   
2               149.0            2.0        2.0        50.0         4.0   
3               149.0            2.0        2.0        50.0         4.0   
4               149.0            2.0        2.0        50.0         4.0   

   num_75_sum  num_985_sum  num_100_sum  num_unq_sum  total_secs_sum  \
0         6.0          5.0        246.0        206.0       64499.977   
1         6.0          5.0        246.0        206.0       64499.977   
2         6.0          5.0        246.0        206.0       64499.977   
3         6.0          5.0        246.0        206.0       64499.977   
4         6.0          5.0        246.0        206.0       64499.977   

   is_weekend_sum  is_weekday_sum  num_25_mean  num_50_mean  num_75_mean  \
0               5               6        4.547       0.3635       0.5454   
1               5               6        4.547       0.3635       0.5454   
2               5               6        4.547       0.3635       0.5454   
3               5               6        4.547       0.3635       0.5454   
4               5               6        4.547       0.3635       0.5454   

   num_985_mean  num_100_mean  num_unq_mean  total_secs_mean  num_25_std  \
0        0.4546         22.36         18.73           5864.0       4.734   
1        0.4546         22.36         18.73           5864.0       4.734   
2        0.4546         22.36         18.73           5864.0       4.734   
3        0.4546         22.36         18.73           5864.0       4.734   
4        0.4546         22.36         18.73           5864.0       4.734   

   num_50_std  num_75_std  num_985_std  num_100_std  num_unq_std  \
0       0.481      0.8906       0.8906        19.42        16.14   
1       0.481      0.8906       0.8906        19.42        16.14   
2       0.481      0.8906       0.8906        19.42        16.14   
3       0.481      0.8906       0.8906        19.42        16.14   
4       0.481      0.8906       0.8906        19.42        16.14   

   total_secs_std  active_days  payment_plan_days_mean  plan_list_price_mean  \
0          4748.0           11                    30.0                 149.0   
1          4748.0           11                    30.0                 149.0   
2          4748.0           11                    30.0                 149.0   
3          4748.0           11                    30.0                 149.0   
4          4748.0           11                    30.0                 149.0   

   actual_amount_paid_mean  transaction_date_count  membership_expire_count  \
0                    149.0                       1                        1   
1                    149.0                       1                        1   
2                    149.0                       1                        1   
3                    149.0                       1                        1   
4               

In [9]:
# saving new dataset
wsdm_dataset.to_csv(r"D:\Work_Masterarbeit\wsdm\preprocessed\prepared_train.csv", index=False)

In [10]:
# reading csv file for removing duplicates
wsdm_dataset = pd.read_csv(r"D:\Work_Masterarbeit\wsdm\preprocessed\prepared_train.csv")
# getting all columns of train dataset
wsdm_dataset.columns

Index(['msno', 'is_churn', 'city', 'bd', 'gender', 'registered_via',
       'payment_method_id', 'payment_plan_days', 'plan_list_price',
       'actual_amount_paid', 'is_auto_renew', 'is_cancel', 'num_25_sum',
       'num_50_sum', 'num_75_sum', 'num_985_sum', 'num_100_sum', 'num_unq_sum',
       'total_secs_sum', 'is_weekend_sum', 'is_weekday_sum', 'num_25_mean',
       'num_50_mean', 'num_75_mean', 'num_985_mean', 'num_100_mean',
       'num_unq_mean', 'total_secs_mean', 'num_25_std', 'num_50_std',
       'num_75_std', 'num_985_std', 'num_100_std', 'num_unq_std',
       'total_secs_std', 'active_days', 'payment_plan_days_mean',
       'plan_list_price_mean', 'actual_amount_paid_mean',
       'transaction_date_count', 'membership_expire_count', 'activity_period',
       'inactive_days', 'is_rare', 'avg_time_perday', 'unq_track_perday',
       'till_25_perday', 'till_50_perday', 'till_75_perday', 'till_985_perday',
       'till_full_perday', 'discount', 'is_discount', 'days_since_last_l

In [11]:
# removing all duplicate rows for a particular id
wsdm_dataset = wsdm_dataset.drop_duplicates('msno', keep='first', inplace=False)
# setting max_columns to 100, so it displays all the columns of train dataset
pd.set_option('display.max_columns', 100)
# displaying head (top 10) rows of train dataset
wsdm_dataset.head(10)

msno  is_churn  city    bd  \
0    ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0   
11   f/NmvEzHfhINFEYZTR05prUdr+E+3+oewvweYz9cCQE=         1  13.0  20.0   
17   zLo9f73nGGT1p21ltZC3ChiRnAVvgibMyazbCxvWPcg=         1  13.0  18.0   
57   8iF/+8HY8lJKFrTc7iR9ZYGCG2Ecrogbc2Vy5YhsfhQ=         1   1.0  28.0   
67   K6fja4+jmoZ5xG6BypqX80Uw/XKpMgrEMdG2edFOxnA=         1  13.0  35.0   
187  ibIHVYBqxGwrSExE63/omeDD99M5vYB3CN2HzkEY+eM=         1  22.0  28.0   
191  kVmM8X4iBPCOfK/m1l3KR7LjFoxD1EcKYCc76F5IAWw=         1   1.0  28.0   
192  moRTKhKIDvb+C8ZHOgmaF4dXMLk0jOn65d7a8tQ2Eds=         1   9.0  28.0   
223  dW/tPZMDh2Oz/ksduEctJbsz0MXw3kay/1AlZCq3EbI=         1  13.0  21.0   
281  otEcMhAX3mU4gumUSogqgteN3oaNmhdmTkoof2iRYEE=         1  13.0  28.0   

     gender  registered_via  payment_method_id  payment_plan_days  \
0       1.0             3.0                0.0               30.0   
11      1.0             3.0               36.0               30.0   
17      1.0             3.0               17.0               30.0   
57      0.0             7.0               41.0               30.0   
67      2.0             7.0               41.0               30.0   
187     0.0             3.0               40.0               30.0   
191     0.0             7.0               41.0               30.0   
192     2.0             3.0               38.0               30.0   
223     2.0             9.0               38.0               30.0   
281     0.0             7.0               40.0               30.0   

     plan_list_price  actual_amount_paid  is_auto_renew  is_cancel  \
0              149.0               149.0            2.0        2.0   
11             180.0               180.0            0.0        0.0   
17               0.0                 0.0            0.0        0.0   
57             149.0               149.0            1.0        0.0   
67              99.0                99.0            1.0        0.0   
187            149.0               149.0            1.0        1.0   
191             99.0                99.0            1.0        1.0   
192            149.0               149.0            0.0        0.0   
223            149.0               149.0            0.0        0.0   
281            149.0               149.0            1.0        0.0   

     num_25_sum  num_50_sum  num_75_sum  num_985_sum  num_100_sum  \
0          50.0         4.0         6.0          5.0        246.0   
11          0.0         4.0         2.0          0.0         26.0   
17        150.0        50.0        38.0         30.0        410.0   
57         20.0         0.0         0.0          0.0        140.0   
67         72.0        56.0        32.0         32.0       3928.0   
187         6.0         1.0         1.0          4.0         37.0   
191         2.0         0.0         0.0          0.0         14.0   
192        78.0        20.0        25.0         16.0        812.0   
223       244.0        46.0        36.0         22.0       1186.0   
281       148.0        28.0        22.0         32.0        730.0   

     num_unq_sum  total_secs_sum  is_weekend_sum  is_weekday_sum  num_25_mean  \
0          206.0       64499.977               5               6        4.547   
11          30.0        6986.509               3               3        0.000   
17         732.0      135620.940              16              24        3.750   
57         160.0       38807.650               0              10        2.000   
67        4384.0      992830.000              24              96        0.600   
187         48.0       10110.665               2               2        1.500   
191         16.0        3880.765               0               1        2.000   
192        649.0      221420.770               8              23        2.516   
223       1118.0      319777.620              14              44        4.207   
281        650.0      195195.080               8              42        2.960   

     num_50_mean  num_75_mean  num_9

In [12]:
# getting shape of train dataset
wsdm_dataset.shape

(970960, 59)

In [13]:
# getting columns and converting them to list
cols = wsdm_dataset.columns.tolist()
# saving them for further use
np.savez_compressed(r"D:\Work_Masterarbeit\wsdm\preprocessed\columns.npz", cols)

In [14]:
# saving actual train dataset
wsdm_dataset.to_csv(r"D:\Work_Masterarbeit\wsdm\preprocessed\actual_train.csv", index=False)

<h2> 3.2 Train Test Split

In [3]:
# loading actual wsdm data
wsdm_data = pd.read_csv(r"D:\Work_Masterarbeit\wsdm\preprocessed\actual_train.csv")
# setting max_columns to 100, so it displays all the columns of train data
pd.set_option('display.max_columns', 100)
# displaying head (top 5) rows of train data
wsdm_data.head()

msno  is_churn  city    bd  gender  \
0  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   
1  f/NmvEzHfhINFEYZTR05prUdr+E+3+oewvweYz9cCQE=         1  13.0  20.0     1.0   
2  zLo9f73nGGT1p21ltZC3ChiRnAVvgibMyazbCxvWPcg=         1  13.0  18.0     1.0   
3  8iF/+8HY8lJKFrTc7iR9ZYGCG2Ecrogbc2Vy5YhsfhQ=         1   1.0  28.0     0.0   
4  K6fja4+jmoZ5xG6BypqX80Uw/XKpMgrEMdG2edFOxnA=         1  13.0  35.0     2.0   

   registered_via  payment_method_id  payment_plan_days  plan_list_price  \
0             3.0                0.0               30.0            149.0   
1             3.0               36.0               30.0            180.0   
2             3.0               17.0               30.0              0.0   
3             7.0               41.0               30.0            149.0   
4             7.0               41.0               30.0             99.0   

   actual_amount_paid  is_auto_renew  is_cancel  num_25_sum  num_50_sum  \
0               149.0            2.0        2.0        50.0         4.0   
1               180.0            0.0        0.0         0.0         4.0   
2                 0.0            0.0        0.0       150.0        50.0   
3               149.0            1.0        0.0        20.0         0.0   
4                99.0            1.0        0.0        72.0        56.0   

   num_75_sum  num_985_sum  num_100_sum  num_unq_sum  total_secs_sum  \
0         6.0          5.0        246.0        206.0       64499.977   
1         2.0          0.0         26.0         30.0        6986.509   
2        38.0         30.0        410.0        732.0      135620.940   
3         0.0          0.0        140.0        160.0       38807.650   
4        32.0         32.0       3928.0       4384.0      992830.000   

   is_weekend_sum  is_weekday_sum  num_25_mean  num_50_mean  num_75_mean  \
0               5               6        4.547       0.3635       0.5454   
1               3               3        0.000       0.6665       0.3333   
2              16              24        3.750       1.2500       0.9500   
3               0              10        2.000       0.0000       0.0000   
4              24              96        0.600       0.4666       0.2666   

   num_985_mean  num_100_mean  num_unq_mean  total_secs_mean  num_25_std  \
0        0.4546        22.360         18.73           5864.0       4.734   
1        0.0000         4.332          5.00           1164.0       0.000   
2        0.7500        10.250         18.30           3390.0       2.980   
3        0.0000        14.000         16.00           3880.0       0.000   
4        0.2666        32.720         36.53           8270.0       0.712   

   num_50_std  num_75_std  num_985_std  num_100_std  num_unq_std  \
0      0.4810      0.8906       0.8906       19.420       16.140   
1      1.1050      0.4714       0.0000        2.926        2.887   
2      1.4795      1.0710       0.7666       12.900       19.250   
3      0.0000      0.0000       0.0000        0.000        0.000   
4      0.8057      0.4421       0.4421       19.060       14.560   

   total_secs_std  active_days  payment_plan_days_mean  plan_list_price_mean  \
0          4748.0           11                    30.0                 149.0   
1           741.5            6                    30.0                 180.0   
2          3882.0           20                    30.0                  74.5   
3             0.0            1                    30.0                 149.0   
4          4612.0           15                    30.0                  99.0   

   actual_amount_paid_mean  transaction_date_count  membership_expire_count  \
0                    149.0                       1                        1   
1                    180.0                       1                        1   
2                     74.5                       2                        2   
3                    149.0                       5                       10   
4               

In [4]:
## Sample 50% of Data due to memory restrictions
wsdm_data = wsdm_data.sample(frac=.5)

In [5]:
# replacing all nan and infinity to 0
wsdm_data = wsdm_data.replace(np.nan, 0, inplace=False)
wsdm_data = wsdm_data.replace(np.inf, 0, inplace=False)

In [6]:
# https://stackoverflow.com/a/29651514
def normalize(df):
    result1 = df.copy()
    for feature_name in df.columns:
        if (str(feature_name) != str('msno') and str(feature_name)!=str('is_churn')):
            max_value = df[feature_name].max()
            min_value = df[feature_name].min()
            result1[feature_name] = (df[feature_name] - min_value) / (max_value - min_value)
    return result1


wsdm_dataset = normalize(wsdm_data)

In [7]:
# getting labels for y
labels = wsdm_dataset['is_churn'].values
labels

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

In [8]:
# getting labels for X
data = wsdm_dataset.drop(['msno', 'is_churn'], axis=1, inplace=False)
data

city        bd  gender  registered_via  payment_method_id  \
643916  0.045455  0.385714     0.0        1.000000           0.731707   
714429  0.681818  0.385714     0.5        0.692308           0.902439   
635595  0.227273  0.542857     1.0        0.692308           0.975610   
196608  0.045455  0.385714     0.0        0.538462           1.000000   
266013  0.590909  0.300000     0.5        0.230769           0.000000   
...          ...       ...     ...             ...                ...   
667437  0.636364  0.385714     0.0        0.230769           0.707317   
369181  0.045455  0.385714     0.0        0.538462           1.000000   
99447   0.045455  0.385714     0.0        0.538462           1.000000   
411984  0.045455  0.385714     0.0        0.538462           1.000000   
878578  0.045455  0.385714     0.0        0.538462           1.000000   

        payment_plan_days  plan_list_price  actual_amount_paid  is_auto_renew  \
643916                1.0         0.827778            0.827778            0.5   
714429                1.0         0.827778            0.827778            0.5   
635595                1.0         0.827778            0.827778            0.5   
196608                1.0         0.550000            0.550000            0.5   
266013                1.0         0.827778            0.827778            1.0   
...                   ...              ...                 ...            ...   
667437                1.0         1.000000            1.000000            0.5   
369181                1.0         0.550000            0.550000            0.5   
99447                 1.0         0.550000            0.550000            0.5   
411984                1.0         0.550000            0.550000            0.5   
878578                1.0         0.550000            0.550000            0.5   

        is_cancel  num_25_sum  num_50_sum  num_75_sum  num_985_sum  \
643916        0.0    0.000107    0.000000    0.000000     0.000000   
714429        0.0    0.005120    0.005285    0.005233     0.010286   
635595        0.0    0.006507    0.007400    0.008140     0.008497   
196608        0.0    0.000107    0.000000    0.000000     0.000000   
266013        1.0    0.000213    0.000000    0.000291     0.000000   
...           ...         ...         ...         ...          ...   
667437        0.0    0.002133    0.001586    0.002035     0.002236   
369181        0.0    0.000107    0.000000    0.000000     0.000000   
99447         0.0    0.000107    0.000000    0.000000     0.000000   
411984        0.0    0.000107    0.000000    0.000000     0.000000   
878578        0.0    0.001493    0.002114    0.001744     0.002683   

        num_100_sum  num_unq_sum  total_secs_sum  is_weekend_sum  \
643916     0.000215     0.000198        0.000230        0.000000   
714429     0.011244     0.006328        0.012486        0.007692   
635595     0.007501     0.007805        0.009188        0.007692   
196608     0.000215     0.000198        0.000230        0.000000   
266013     0.000353     0.000330        0.000317        0.000000   
...             ...          ...             ...             ...   
667437     0.009756     0.007726        0.009948        0.004808   
369181     0.000215     0.000198        0.000230        0.000000   
99447      0.000215     0.000198        0.000230        0.000000   
411984     0.000215     0.000198        0.000230        0.000000   
878578     0.003605     0.002848        0.003295        0.001923   

        is_weekday_sum  num_25_mean  num_50_mean  num_75_mean  num_985_mean  \
643916        0.000323     0.133333     0.000000     0.000000      0.000000   
714429        0.007429     0.206533     0.161250     0.193533      0.247233   
635595        0.007429     0.262400     0.225825     0.301100      0.204333   
196608        0.000323     0.133333     0.000000     0.000000      0.000000   
266013        0.000646     0.133333     0.000000     0.166667      0.000000   
...                ...        

In [9]:
# train test split
X_train, X_test, y_train, y_test = train_test_split(data, labels, test_size=0.3, random_state=42)

print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(339836, 57) (339836,)
(145644, 57) (145644,)


In [10]:
X_train

city        bd  gender  registered_via  payment_method_id  \
150239  0.045455  0.385714     0.0        0.538462           1.000000   
71000   0.045455  0.385714     0.0        0.538462           1.000000   
218055  0.045455  0.385714     0.0        0.538462           1.000000   
150218  0.045455  0.385714     0.0        0.230769           0.926829   
151405  0.045455  0.385714     0.0        0.538462           1.000000   
...          ...       ...     ...             ...                ...   
345905  0.045455  0.385714     0.0        0.538462           1.000000   
517244  0.227273  0.285714     0.5        0.538462           1.000000   
740938  0.045455  0.385714     0.0        0.538462           1.000000   
727070  0.000000  0.385714     0.0        0.000000           1.000000   
471544  0.636364  0.700000     1.0        0.692308           0.829268   

        payment_plan_days  plan_list_price  actual_amount_paid  is_auto_renew  \
150239                1.0         0.550000            0.550000            0.5   
71000                 1.0         0.827778            0.827778            0.5   
218055                1.0         0.550000            0.550000            0.5   
150218                1.0         0.827778            0.827778            0.0   
151405                1.0         0.550000            0.550000            0.5   
...                   ...              ...                 ...            ...   
345905                1.0         0.550000            0.550000            0.5   
517244                1.0         0.827778            0.827778            0.5   
740938                1.0         0.555556            0.555556            0.5   
727070                1.0         0.550000            0.550000            0.5   
471544                1.0         0.827778            0.827778            0.5   

        is_cancel  num_25_sum  num_50_sum  num_75_sum  num_985_sum  \
150239        0.0    0.003147    0.003436    0.002907     0.005814   
71000         0.0    0.003040    0.005285    0.002616     0.005367   
218055        0.0    0.000533    0.000793    0.000872     0.003131   
150218        0.0    0.007680    0.009514    0.008721     0.012075   
151405        0.0    0.001120    0.002114    0.000872     0.001789   
...           ...         ...         ...         ...          ...   
345905        0.0    0.000107    0.000000    0.000000     0.000000   
517244        0.5    0.008000    0.006342    0.012209     0.024150   
740938        0.0    0.002827    0.002114    0.002616     0.003578   
727070        0.0    0.000107    0.000000    0.000000     0.000000   
471544        0.0    0.001653    0.002114    0.002035     0.004472   

        num_100_sum  num_unq_sum  total_secs_sum  is_weekend_sum  \
150239     0.002915     0.004509        0.003671        0.002885   
71000      0.003666     0.003718        0.003710        0.005769   
218055     0.000399     0.000567        0.000508        0.001923   
150218     0.010738     0.011404        0.011410        0.007692   
151405     0.000184     0.000765        0.000330        0.001923   
...             ...          ...             ...             ...   
345905     0.000215     0.000198        0.000230        0.000000   
517244     0.015739     0.012801        0.015194        0.013462   
740938     0.002792     0.002742        0.002928        0.002885   
727070     0.000215     0.000198        0.000230        0.000000   
471544     0.008989     0.007370        0.008351        0.006731   

        is_weekday_sum  num_25_mean  num_50_mean  num_75_mean  num_985_mean  \
150239        0.004845     0.218467     0.180500     0.185233      0.240667   
71000         0.006137     0.152000     0.200000     0.120000      0.160000   
218055        0.001615     0.095267     0.107125     0.142833      0.333333   
150218        0.006460     0.343000     0.321500     0.357000      0.321467   
151405        0.002261     0.155600     0.222175     0.111100      0.148100   
...                ...        

In [11]:
# converting X_train to numpy array
wsdm_arr = X_train.values
wsdm_arr

array([[0.04545455, 0.38571429, 0.        , ..., 0.55      , 0.01221001,
        0.69305027],
       [0.04545455, 0.38571429, 0.        , ..., 0.82777778, 0.01465201,
        0.69911926],
       [0.04545455, 0.38571429, 0.        , ..., 0.55      , 0.        ,
        0.92154148],
       ...,
       [0.04545455, 0.38571429, 0.        , ..., 0.55555556, 0.03174603,
        0.9990385 ],
       [0.        , 0.38571429, 0.        , ..., 0.55      , 0.02930403,
        0.84518288],
       [0.63636364, 0.7       , 1.        , ..., 0.82777778, 0.        ,
        0.30616515]])

In [12]:
# checking shape of train array
wsdm_arr.shape

(339836, 57)

In [13]:
# saving it for further use
np.savez_compressed(r"D:\Work_Masterarbeit\wsdm\preprocessed\actual_train.npz", wsdm_arr)

In [14]:
# displaying y_train
wsdm_labels_arr = y_train
wsdm_labels_arr

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

In [15]:
# checking shape of train labels
wsdm_labels_arr.shape

(339836,)

In [16]:
# saving it for further use
np.savez_compressed(r"D:\Work_Masterarbeit\wsdm\preprocessed\actual_wsdm_labels.npz", wsdm_labels_arr)

In [17]:
# converting X_test to numpy array
cv_arr = X_test.values
cv_arr

array([[0.04545455, 0.38571429, 0.        , ..., 0.55      , 0.02075702,
        0.92448752],
       [0.63636364, 0.28571429, 1.        , ..., 0.55      , 0.004884  ,
        0.85054421],
       [0.04545455, 0.38571429, 0.        , ..., 0.71666667, 0.01709402,
        0.53684089],
       ...,
       [0.22727273, 0.32857143, 0.5       , ..., 0.82777778, 0.        ,
        0.69679628],
       [0.04545455, 0.38571429, 0.        , ..., 0.82777778, 0.01098901,
        0.62226068],
       [0.27272727, 0.3       , 0.5       , ..., 0.82777778, 0.        ,
        0.69221953]])

In [18]:
# checking shape of cv array
cv_arr.shape

(145644, 57)

In [19]:
# saving it for further use
np.savez_compressed(r"D:\Work_Masterarbeit\wsdm\preprocessed\actual_cv.npz", cv_arr)

In [20]:
# displaying y_test
cv_labels_arr = y_test
cv_labels_arr

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

In [21]:
# checking shape of cv labels array
cv_labels_arr.shape

(145644,)

In [22]:
# saving it for further use
np.savez_compressed(r"D:\Work_Masterarbeit\wsdm\preprocessed\actual_cv_labels.npz", cv_labels_arr)

<h1> 5. Modelling

In [23]:
# loading library
import warnings
warnings.filterwarnings("ignore")

import numpy as np
import pandas as pd
import joblib
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import RandomizedSearchCV
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import log_loss

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
import xgboost as xgb


In [24]:
# reference: https://machinelearningmastery.com/how-to-save-a-numpy-array-to-file-for-machine-learning/

X_train = np.load(r"D:\Work_Masterarbeit\wsdm\preprocessed\actual_train.npz")
y_train = np.load(r"D:\Work_Masterarbeit\wsdm\preprocessed\actual_wsdm_labels.npz")
X_test = np.load(r"D:\Work_Masterarbeit\wsdm\preprocessed\actual_cv.npz")
y_test = np.load(r"D:\Work_Masterarbeit\wsdm\preprocessed\actual_cv_labels.npz")


In [25]:
X_train = X_train['arr_0']
X_test = X_test['arr_0']
y_train = y_train['arr_0']
y_test = y_test['arr_0']

# getting shapes
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(339836, 57) (339836,)
(145644, 57) (145644,)


In [26]:
# loading columns
cols = np.load(r"D:\Work_Masterarbeit\wsdm\preprocessed\columns.npz")
# converting numpy array to python list
cols = cols['arr_0'].tolist()
# removing first two cols 'msno' and 'is_churn'
cols = cols[2:]

<h2> Logistic Regression

In [27]:
# logistic regression model
lr_model = LogisticRegression()

params = {'C' : [0.0001,0.001,0.01,0.1,1,10,100]}

# Random search cv
rs_lr = RandomizedSearchCV(lr_model, param_distributions=params, verbose=10, n_jobs=-1)
rs_lr.fit(X_train, y_train)
# getting best parameters
rs_lr.best_params_

Fitting 5 folds for each of 7 candidates, totalling 35 fits


{'C': 100}

In [28]:
%%time

# fitting logistic regression model with best parameter values
lr_model = LogisticRegression(penalty='l2', tol=0.0001, C=10, solver='newton-cg', class_weight='balanced', n_jobs=-1)
lr_model.fit(X_train, y_train)

Wall time: 1min 35s


LogisticRegression(C=10, class_weight='balanced', n_jobs=-1, solver='newton-cg')

In [29]:
# Reference: https://machinelearningmastery.com/save-load-machine-learning-models-python-scikit-learn/#:~:text=Saving%20Your%20Model-,Save%20Your%20Model%20with%20pickle,it%20to%20make%20new%20predictions.
# saving the trained logistic regression ML model
filename = r"D:\Work_Masterarbeit\wsdm\preprocessed\finalized_model_lr.sav"
joblib.dump(lr_model, filename)

['D:\\Work_Masterarbeit\\wsdm\\preprocessed\\finalized_model_lr.sav']

In [30]:
# loading the best model from disk
filename = r"D:\Work_Masterarbeit\wsdm\preprocessed\finalized_model_lr.sav"
loaded_model = joblib.load(filename)

<h3> Model Evaluation

In [31]:
from sklearn.metrics import classification_report,confusion_matrix
from sklearn.metrics import f1_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score
from sklearn.metrics import accuracy_score

lr_predict=lr_model.predict(X_test)
lr_predict


print('Accuracy:', accuracy_score(y_test, lr_predict))
print('F1 score weighted:', f1_score(y_test, lr_predict,average='weighted'))
print('F1 score macro:', f1_score(y_test, lr_predict, average='macro'))
print('Recall:', recall_score(y_test, lr_predict,average='weighted'))
print('Precision:', precision_score(y_test, lr_predict,average='weighted'))
print('\n clasification report:\n', classification_report(y_test, lr_predict))
print('\n confussion matrix:\n',confusion_matrix(y_test, lr_predict))

Accuracy: 0.9203056768558951
F1 score weighted: 0.9286840357035214
F1 score macro: 0.8110516012623766
Recall: 0.9203056768558951
Precision: 0.9465875485616565

 clasification report:
               precision    recall  f1-score   support

           0       0.99      0.92      0.95    132443
           1       0.54      0.88      0.67     13201

    accuracy                           0.92    145644
   macro avg       0.76      0.90      0.81    145644
weighted avg       0.95      0.92      0.93    145644


 confussion matrix:
 [[122393  10050]
 [  1557  11644]]


In [32]:
## Logloss
# predicting probabilities for X_cv
pred_test = loaded_model.predict_proba(X_test)
# getting probabilities corresponding to class label 1 only
pred_test = pred_test[:,1]
# calculating log loss
print('Test log-loss: ', log_loss(y_test, pred_test))

Test log-loss:  0.2901594007798089


<h2> Decision Tree

In [34]:
%%time

# decision tree classifier model
dtc_model = DecisionTreeClassifier()

params = {'max_depth' : [2,3,5,7,10,15],
          'min_samples_split' : [2,3,5,7,10],
          'min_samples_leaf' : [1,2,3],
          'criterion' : ['gini', 'entropy'],
          'splitter' : ['best', 'random']}

# Random search cv
rs_dtc = RandomizedSearchCV(dtc_model, param_distributions=params, verbose=10, n_jobs=-1)
rs_dtc.fit(X_train, y_train)
# getting best parameters
rs_dtc.best_params_

Fitting 5 folds for each of 10 candidates, totalling 50 fits
Wall time: 38.7 s


{'splitter': 'best',
 'min_samples_split': 2,
 'min_samples_leaf': 2,
 'max_depth': 15,
 'criterion': 'gini'}

In [35]:
%%time

# training a decision tree classifier model with best parameters
dtc_model = DecisionTreeClassifier(criterion='gini', splitter='best', max_depth=5, min_samples_split=3, min_samples_leaf=1)
dtc_model.fit(X_train, y_train)

Wall time: 3.02 s


DecisionTreeClassifier(max_depth=5, min_samples_split=3)

In [36]:
# https://machinelearningmastery.com/save-load-machine-learning-models-python-scikit-learn/#:~:text=Saving%20Your%20Model-,Save%20Your%20Model%20with%20pickle,it%20to%20make%20new%20predictions.
# saving the decision tree classifier model
filename = r"D:\Work_Masterarbeit\wsdm\preprocessed\finalized_model_dtc.sav"
joblib.dump(dtc_model, filename)

['D:\\Work_Masterarbeit\\wsdm\\preprocessed\\finalized_model_dtc.sav']

In [37]:
# loading the best decision tree classifier model from disk
filename = r"D:\Work_Masterarbeit\wsdm\preprocessed\finalized_model_dtc.sav"
loaded_model = joblib.load(filename)

<h3> Model Evaluation

In [38]:
from sklearn.metrics import classification_report,confusion_matrix
from sklearn.metrics import f1_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score
from sklearn.metrics import accuracy_score

dtc_modelpred=dtc_model.predict(X_test)
dtc_modelpred


print('Accuracy:', accuracy_score(y_test, dtc_modelpred))
print('F1 score weighted:', f1_score(y_test, dtc_modelpred,average='weighted'))
print('F1 score macro:', f1_score(y_test, dtc_modelpred, average='macro'))
print('Recall:', recall_score(y_test, dtc_modelpred,average='weighted'))
print('Precision:', precision_score(y_test, dtc_modelpred,average='weighted'))
print('\n clasification report:\n', classification_report(y_test, dtc_modelpred))
print('\n confussion matrix:\n',confusion_matrix(y_test, dtc_modelpred))

Accuracy: 0.9696039658345006
F1 score weighted: 0.9686350892116973
F1 score macro: 0.9019698922951905
Recall: 0.9696039658345006
Precision: 0.9684794940520802

 clasification report:
               precision    recall  f1-score   support

           0       0.98      0.99      0.98    132443
           1       0.88      0.77      0.82     13201

    accuracy                           0.97    145644
   macro avg       0.93      0.88      0.90    145644
weighted avg       0.97      0.97      0.97    145644


 confussion matrix:
 [[131096   1347]
 [  3080  10121]]


In [40]:
## Logloss
# predicting probabilities for X_cv
pred_test = loaded_model.predict_proba(X_test)
# getting probabilities corresponding to class label 1 only
pred_test = pred_test[:,1]
# calculating log loss
print('Test log-loss: ', log_loss(y_test, pred_test))

Test log-loss:  0.10959740889361416


<h2> Random Forest

In [41]:
%%time

# training the random forest classifier for best hyper parameters
rf_model = RandomForestClassifier(n_estimators=500, criterion='gini', max_depth=15, min_samples_split=3, min_samples_leaf=2, max_features='sqrt', bootstrap=False, n_jobs=-1)
rf_model.fit(X_train, y_train)

Wall time: 1min 21s


RandomForestClassifier(bootstrap=False, max_depth=15, max_features='sqrt',
                       min_samples_leaf=2, min_samples_split=3,
                       n_estimators=500, n_jobs=-1)

In [42]:
# https://machinelearningmastery.com/save-load-machine-learning-models-python-scikit-learn/#:~:text=Saving%20Your%20Model-,Save%20Your%20Model%20with%20pickle,it%20to%20make%20new%20predictions.
# saving the decision tree classifier model
filename = r"D:\Work_Masterarbeit\wsdm\preprocessed\finalized_model_rf.sav"
joblib.dump(rf_model, filename)

['D:\\Work_Masterarbeit\\wsdm\\preprocessed\\finalized_model_rf.sav']

In [43]:
# loading the best model from disk
filename = r"D:\Work_Masterarbeit\wsdm\preprocessed\finalized_model_rf.sav"
loaded_model = joblib.load(filename)

<h3>Model Evaluation

In [46]:
from sklearn.metrics import classification_report,confusion_matrix
from sklearn.metrics import f1_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score
from sklearn.metrics import accuracy_score

rf_modelpred=rf_model.predict(X_test)
rf_modelpred


print('Accuracy:', accuracy_score(y_test, rf_modelpred))
print('F1 score weighted:', f1_score(y_test, rf_modelpred,average='weighted'))
print('F1 score macro:', f1_score(y_test, rf_modelpred, average='macro'))
print('Recall:', recall_score(y_test, rf_modelpred,average='weighted'))
print('Precision:', precision_score(y_test, rf_modelpred,average='weighted'))
print('\n clasification report:\n', classification_report(y_test, rf_modelpred))
print('\n confussion matrix:\n',confusion_matrix(y_test, rf_modelpred))

Accuracy: 0.974932026036088
F1 score weighted: 0.9744936554809953
F1 score macro: 0.9213174787461371
Recall: 0.974932026036088
Precision: 0.9743123174333281

 clasification report:
               precision    recall  f1-score   support

           0       0.98      0.99      0.99    132443
           1       0.89      0.82      0.86     13201

    accuracy                           0.97    145644
   macro avg       0.94      0.91      0.92    145644
weighted avg       0.97      0.97      0.97    145644


 confussion matrix:
 [[131109   1334]
 [  2317  10884]]


In [47]:
## Logloss
# predicting probabilities for X_cv
pred_test = loaded_model.predict_proba(X_test)
# getting probabilities corresponding to class label 1 only
pred_test = pred_test[:,1]
# calculating log loss
print('Test log-loss: ', log_loss(y_test, pred_test))

Test log-loss:  0.07355619805123177
